src/compiler/checker.ts

```diff
@@ -23628,6 +23628,20 @@ namespace ts {
             }
         }
 
+        function statementRefersToSuperOrThis(statement: Statement): boolean {
+            return !!forEachChild(statement, function visitNode(node: Node): true | undefined {
+                if (node.kind === SyntaxKind.SuperKeyword || node.kind === SyntaxKind.ThisKeyword) {
+                    return true;
+                }
+
+                if (isClassLike(node) || isFunctionLike(node)) {

```

Comment:
isFunctionLike includes immediately executed arrow functions or object literal methods/accessors with computed names, that could use `this`

Reply:
Added a new `isNewThisScope` method - I think that should cover it?